src/service/ingestion/mod.rs

```diff
@@ -368,7 +368,18 @@ pub fn apply_stream_functions(
             value = apply_vrl_fn(runtime, vrl_runtime, &value, org_id, stream_name);
         }
     }
-    flatten::flatten_with_level(value, get_config().limit.ingest_flatten_level)
+
+    let flatten_level = get_config().limit.ingest_flatten_level;
+
+    if let Value::Array(v) = value {
+        return v
+            .into_iter()
+            .map(|item| flatten::flatten_with_level(item, flatten_level))
+            .collect::<Result<Vec<_>, _>>()
+            .map(|v| Value::Array(v));
+    }
+
+    flatten::flatten_with_level(value, flatten_level)
```

Comment:
It is better to return a fixed type from here method as the code that handles the return value is extremely awkward. 
The only possible return types from here are : `Map<String, Value>` or `Vec<Map<String, Value>>`.

Any code that handles this return value will have to check `Value` for all combinations again.
For eg:
```rust
let local_vals = match apply_stream_functions(...) {
    json::Value::Object(val) => vec![val],
    json::Value::Array(v) => {
        v.into_iter().filter_map(|item| {
            if let json::Value::Object(map) = item {
                Some(map)
            } else {
                None // this will never be entered as apply_stream_functions will fail if the item is not an object
            }
        }).collect()
    }
    _ => unreachable!(),
};
```

Since the intent from this function is to always return valid records it makes sense to code this into the return type. I would suggest something like this:

```rust
pub enum RecordType {
    Single(Map<String, Value>),
    Multiple(Vec<Map<String, Value>>),
}
```
This greatly simplifies the handling of the same code:
```rust
fn apply_stream_functions(...) -> Result<RecordType> {
    // ...
}
// ...
match apply_stream_functions(...)? {
    Single(map) => // ...,
    Multiple(v) => //..., each item is a map, no need for filtering
}
```

Reply:
To limit the scope of changes, I have opted to write down my thoughts here rather than implementing this.